### PR TITLE
add postgres support to storj-sim

### PR DESCRIPTION
**Pull request by JessicaGreben**
What: 
Add the ability to run storj-sim with postgres (in addition to the current default sqlite). This requires to add support for schemas in postgres code for the satellite master db and also the pointer/metainfo db.  This PR also adds an env var to test-sim.sh script to use postgres.

Examples:
```
$ storj-sim network --postgres=postgres://postgres:<pw>@localhost/testdb?sslmode=disable setup

$ storj-sim network run
```

```
$ export STORJ_SIM_POSTGRES=postgres://postgres:<pw>@localhost/testdb?sslmode=disable

$ make test-sim
```

Why:
So that we can test with postgres since that is whats used in production also so that there is consistent support for the same dbs across our test frameworks.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: cmd/storj-sim/main.go [hunk 1]
```diff
@@ -23,6 +23,11 @@ type Flags struct {
 	IsDev bool
 
 	OnlyEnv bool // only do things necessary for loading env vars
+
+	// Connection string for the satellite masterdb
+	SatelliteDB string
```
Comment (egonelbre) [inline cmd/storj-sim/main.go]: A simpler design would allow to specify `--postgres postgres://storj:storj-pass@localhost/teststorj?sslmode=disable` and rest happens automatically.

File: cmd/storj-sim/network.go [hunk 2]
```diff
@@ -260,6 +260,10 @@ func newNetwork(flags *Flags) (*Processes, error) {
 
 		consoleAuthToken := "secure_token"
 
+		satelliteDB := flags.SatelliteDB
+		if os.Getenv("STORJ_POSTGRES_TEST") != "" {
+			satelliteDB = os.Getenv("STORJ_POSTGRES_TEST")
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: This is for different purpose and probably shouldn't override storj-sim behavior.

File: cmd/storj-sim/network.go [hunk 3]
```diff
@@ -281,6 +285,9 @@ func newNetwork(flags *Flags) (*Processes, error) {
 
 				"--version.server-address", fmt.Sprintf("http://%s/", versioncontrol.Address),
 				"--debug.addr", net.JoinHostPort("127.0.0.1", port(satellitePeer, i, debugHTTP)),
+
+				"--database", satelliteDB,
+				"--metainfo.database-url", flags.MetainfoDB,
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: These won't work with multiple satellites, you need to use schemas to make it work.

File: cmd/storj-sim/network.go (hunk 2 above)
Comment (egonelbre) [reply cmd/storj-sim/network.go]: Also for command line flags the usual behavior is that command line flags override environment variables not the other way around.

File: cmd/satellite/main.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 // Satellite defines satellite configuration
 type Satellite struct {
 	Database string `help:"satellite database connection string" default:"sqlite3://$CONFDIR/master.db"`
+	Schema   string `help:"satellite database schema name" default:""`
```
Comment (egonelbre) [inline cmd/satellite/main.go]: You should not need this. You can use connstr to pass it in.

File: cmd/storj-sim/network.go [hunk 5]
```diff
@@ -281,6 +288,10 @@ func newNetwork(flags *Flags) (*Processes, error) {
 
 				"--version.server-address", fmt.Sprintf("http://%s/", versioncontrol.Address),
 				"--debug.addr", net.JoinHostPort("127.0.0.1", port(satellitePeer, i, debugHTTP)),
+
+				"--database", database,
+				"--schema", schema,
+				// TODO(jg): add connection to postgres for metainfo.databaseURL once metainfo refactor is complete
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: You can add the pointerdb/metainfo url already now, there shouldn't be issues with it. If there are then these should be fixed.

File: cmd/satellite/main.go (hunk 4 above)
Comment (JessicaGreben) [reply cmd/satellite/main.go]: if i dont pass in the schema, then i need to parse it out of the connstr before I can call `db.CreateSchema`.  I thought passing this in was cleaner. thoughts?

File: cmd/storj-sim/network.go (hunk 5 above)
Comment (JessicaGreben) [reply cmd/storj-sim/network.go]: yeah i was having trouble with metainfodb using postgres, but let me try again. I could have been doing something wrong.

File: cmd/satellite/main.go (hunk 4 above)
Comment (egonelbre) [reply cmd/satellite/main.go]: Do you need to explicitly create the schema? But, I would rather handle it transparently without an extra flag.

Comment (JessicaGreben) [reply cmd/satellite/main.go]: I think I do need to create the schema explicitly. At least it was erroring before I added did. Ok let me try to change it to not need an extra flag.

File: cmd/satellite/main.go [hunk 6]
```diff
@@ -141,6 +142,13 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		zap.S().Error("Failed to initialize telemetry batcher: ", err)
 	}
 
+	if runCfg.Schema != "" {
+		err = db.CreateSchema(runCfg.Schema)
```
Comment (egonelbre) [inline cmd/satellite/main.go]: Move this into `satellitedb` `CreateTables` method.

File: satellite/console/consoleweb/server.go [hunk 7]
```diff
@@ -76,7 +77,7 @@ func NewServer(logger *zap.Logger, config Config, service *console.Service, mail
 		mailService: mailService,
 	}
 
-	logger.Debug("Starting Satellite UI...")
+	logger.Debug(fmt.Sprintf("Starting Satellite UI on %s...", server.listener.Addr().String()))
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: `logger.Sugar().Debugf`

File: cmd/satellite/main.go (hunk 6 above)
Comment (JessicaGreben) [reply cmd/satellite/main.go]: I tested out adding `CreateSchema` to the `CreateTables` method, but [`CreateTables`](https://github.com/storj/storj/blob/master/satellite/satellitedb/migrate.go#L22) doesn't have the database connstr available to it.  So I'm not sure what is best.  What I initially chose to do is [this](https://github.com/storj/storj/pull/1908/commits/eb0fb725b99132b29c47dab6b8c37d0362938f6f#diff-e7e4cbc943e802fe544ebbf762a3a459R145) just keep the CreateSchema in this file.

Comment (egonelbre) [reply cmd/satellite/main.go]: You can save the schema or connstr in https://github.com/storj/storj/blob/master/satellite/satellitedb/database.go#L47.

File: internal/dbutil/pgutil/schema.go [hunk 8]
```diff
@@ -28,6 +28,15 @@ func ConnstrWithSchema(connstr, schema string) string {
 	return connstr + "&search_path=" + url.QueryEscape(schema)
 }
 
+// ParseSchemaFromConnstr returns the schema name parsed from the connection string
+func ParseSchemaFromConnstr(connstr string) (string, error) {
+	connstrParts := strings.Split(connstr, "&search_path=")
```
Comment (egonelbre) [inline internal/dbutil/pgutil/schema.go]: There's `url.Parse` which allows to properly decode the string. The `search_path` isn't necessarily the last argument.

Comment (JessicaGreben) [reply internal/dbutil/pgutil/schema.go]: good point

File: cmd/storj-sim/network.go [hunk 9]
```diff
@@ -262,10 +262,13 @@ func newNetwork(flags *Flags) (*Processes, error) {
 		consoleAuthToken := "secure_token"
 
 		var schema string
-		database := "sqlite3://" + process.Directory + "/master.db"
+		sqlDatabase := "sqlite3://" + process.Directory + "/master.db"
+		kvDatabase := "bolt://" + process.Directory + "pointers.db"
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: You might be able to avoid this, if you add the argument dynamically when needed.

Comment (JessicaGreben) [reply cmd/storj-sim/network.go]: good idea, let me try that

File: storage/postgreskv/schema/migrate.go [hunk 10]
```diff
@@ -37,3 +53,11 @@ func PrepareDB(db *sql.DB) error {
 	}
 	return err
 }
+
+func createSchema(db *sql.DB, schema string) error {
+	_, err := db.Exec(`create schema if not exists ` + strconv.QuoteToASCII(schema) + `;`)
+	if err != nil {
```
Comment (mniewrzal) [inline storage/postgreskv/schema/migrate.go]: nit: it can be just `return err`

File: cmd/satellite/main.go [hunk 11]
```diff
@@ -116,7 +116,7 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		zap.S().Fatal(err)
 	}
 
-	db, err := satellitedb.New(log.Named("db"), runCfg.Database)
+	db, err := satellitedb.New(log.Named("satellite master db"), runCfg.Database)
```
Comment (egonelbre) [inline cmd/satellite/main.go]: Keep this as is, since it comes from satellite log, it should be obvious that it's master db.

File: storage/postgreskv/schema/migrate.go [hunk 12]
```diff
@@ -7,22 +7,38 @@ package schema
...
 	if err != nil {
 		return err
 	}
+
+	schema, err := pgutil.ParseSchemaFromConnstr(dbURL)
+	if err != nil {
+		return errs.New("error parsing schema: %+v", err)
+	}
+	if schema != "" {
+		err := createSchema(db, schema)
```
Comment (egonelbre) [inline storage/postgreskv/schema/migrate.go]: There is `pgutil.CreateSchema`

Comment (JessicaGreben) [reply storage/postgreskv/schema/migrate.go]: oops you are right, i missed that. thanks

File: cmd/storj-sim/network.go [hunk 13]
```diff
@@ -284,6 +285,13 @@ func newNetwork(flags *Flags) (*Processes, error) {
 			},
 			"run": {},
 		})
+		var schema string
+		if flags.Postgres != "" {
+			schema = fmt.Sprintf("satellite%d", i)
+			postgresConnstr := pgutil.ConnstrWithSchema(flags.Postgres, schema)
+			dbArgs := []string{"--database", postgresConnstr, "--metainfo.database-url", postgresConnstr}
```
Comment (egonelbre) [inline cmd/storj-sim/network.go]: main database and metainfo probably should use different schemas.
e.g. `satellite/0` and `satellite/0/metainfo`

Comment (JessicaGreben) [reply cmd/storj-sim/network.go]: good point

Comment (egonelbre) [reply cmd/storj-sim/network.go]: They may have some conflicts due to separate initialization. Might be better to leave them separate.

Comment (JessicaGreben) [reply cmd/storj-sim/network.go]: what part of the initialization will conflict?  the createSchema function executes `create schema if not exists` so that should be ok. what else would conflict?

Comment (egonelbre) [reply cmd/storj-sim/network.go]: There are other tables that the postgreskv makes and satellite. I'm unsure whether they conflict... easier it's to use different schemas to ensure it definitely doesn't happen. However, if it definitely doesn't conflict, then I'm fine with keeping them in the same schema.
